docs: updates go.md and go-built-in.md as part of issue 9272 (#9688)
tendermint/tendermint: docs/tutorials/go.md

for state-machine replication. The replicated state-machine, or "application", can be written
in any language that can send and receive protocol buffer messages in a client-server model.
Applications written in Go can also use Tendermint as a library and run the service in the same 
process as the application. 

By following along this tutorial you will create a Tendermint Core application called kvstore, 
a (very) simple distributed BFT key-value store.
The application will be written in Go and  
some understanding of the Go programming language is expected.
If you have never written Go, you may want to go through [Learn X in Y minutes
Where X=Go](https://learnxinyminutes.com/docs/go/) first, to familiarize
yourself with the syntax.

Note: Please use the latest released version of this guide and of Tendermint.
We strongly advise against using unreleased commits for your development.

### Built-in app vs external app
On the one hand, to get maximum performance you can run your application in 
the same process as the Tendermint Core, as long as your application is written in Go. 
[Cosmos SDK](https://github.com/cosmos/cosmos-sdk) is written
this way. 

> If that is the way you with to proceed, use the [Writing a built-in Tendermint Core application in Go](./go-built-in.md) guide instead of this one.

Author: ```suggestion
If that is the way you wish to proceed, use the [Writing a built-in Tendermint Core application in Go](./go-built-in.md) guide instead of this one.
```

> If that is the way you with to proceed, use the [Writing a built-in Tendermint Core application in Go](./go-built-in.md) guide instead of this one.

Author: ```suggestion
If that is the way you with to proceed, use the [Creating a built-in application in Go](./go-built-in.md) guide instead of this one.
```

```bash
mkdir kvstore
```

Inside the example directory, create a `main.go` file with the following content:

```go
package main

import (
    "fmt"
)

func main() {
    fmt.Println("Hello, Tendermint Core")
}
```

When run, this should print "Hello, Tendermint Core" to the standard output.

```bash
cd kvstore
$ go run main.go
Hello, Tendermint Core
```

We are going to use [Go modules](https://github.com/golang/go/wiki/Modules) for

> dependency management, so let's start by including a dependency on the latest

Author: ```suggestion
dependency management, so let's start by including a dependency on the latest version
```

	return abcitypes.ResponseCommit{}
}

func (app *KVStoreApplication) ListSnapshots(snapshots abcitypes.RequestListSnapshots) abcitypes.ResponseListSnapshots {
	return abcitypes.ResponseListSnapshots{}
}

func (app *KVStoreApplication) OfferSnapshot(snapshot abcitypes.RequestOfferSnapshot) abcitypes.ResponseOfferSnapshot {
	return abcitypes.ResponseOfferSnapshot{}
}

func (app *KVStoreApplication) LoadSnapshotChunk(chunk abcitypes.RequestLoadSnapshotChunk) abcitypes.ResponseLoadSnapshotChunk {
	return abcitypes.ResponseLoadSnapshotChunk{}
}

func (app *KVStoreApplication) ApplySnapshotChunk(chunk abcitypes.RequestApplySnapshotChunk) abcitypes.ResponseApplySnapshotChunk {
	return abcitypes.ResponseApplySnapshotChunk{}
}
```

The types used here are defined in the Tendermint library and were added as a dependency
to the project when you ran `go get`. If your IDE is not recognizing the types, go ahead and run the command again.

```bash
go get github.com/tendermint/tendermint@latest
```

Now go back to the `main.go` and modify the `main` function so it matches the following, 
where an instance of the `KVStoreApplication` type is created.

> !!!!!!!!!!!

Reviewer: ?

Author: ```suggestion
```

```go
func main() {
    fmt.Println("Hello, Tendermint Core")

    _ = NewKVStoreApplication()
}
```

You can recompile and run the application now by running `go get` and `go build`, but it does
not do anything.
So let's revisit the code adding the logic needed to implement our minimal key/value store
and to start it along with the Tendermint Service.

### 1.3.1 Add a persistent data store
Our application will need to write its state out to persistent storage so that it
can stop and start without losing all of its data.

For this tutorial, we will use [BadgerDB](https://github.com/dgraph-io/badger), a
a fast embedded key-value store. 

First, add Badger as a dependency of your go module using the `go get` command:

`go get github.com/dgraph-io/badger/v3`

Next, let's update the application and its constructor to receive a handle to the database, as follows: 

```go
type KVStoreApplication struct {
	db           *badger.DB
	onGoingBlock *badger.Txn
}

var _ abcitypes.Application = (*KVStoreApplication)(nil)

func NewKVStoreApplication(db *badger.DB) *KVStoreApplication {
	return &KVStoreApplication{db: db}
}
```

> The `onGoingBlock` keeps track of the transaction that will update the application's state when a block 

Reviewer: ```suggestion
The `onGoingBlock` keeps track of the transactions that will update the application's state when a block 
```

	code := app.isValid(req.Tx)
	return abcitypes.ResponseCheckTx{Code: code}
}
```

While this `CheckTx` is simple and only validates that the transaction is well-formed, 
it is very common for `CheckTx` to make more complex use of the state of an application.
For example, you may refuse to overwrite an existing value, or you can associate 
versions to the key/value pairs and allow the caller to specify a version to
perform a conditional update.

Depending on the checks and on the conditions violated, the function may return
different values, but any response with a non-zero code will be considered invalid 
by Tendermint. Our `CheckTx` logic returns 0 to Tendermint when a transaction passes 
its validation checks. The specific value of the code is meaningless to Tendermint. 
Non-zero codes are logged by Tendermint so applications can provide more specific 
information on why the transaction was rejected.

Note that `CheckTx` does not execute the transaction, it only verifies that that the transaction could be executed. We do not know yet if the rest of the network has agreed to accept this transaction into a block.

> Finally, make sure to add the bytes package to the your `import` stanza at the top of `app.go`:

Reviewer: ```suggestion
Finally, make sure to add the bytes package to your `import` stanza at the top of `app.go`:
```

Author: ```suggestion
Finally, make sure to add the bytes package to the `import` stanza at the top of `app.go`:
```

confuse them with the transactions included in the blocks delivered by Tendermint,
the _application transactions_.

First, let's create a new Badger transaction during `BeginBlock`. All application transactions in the
current block will be executed within this Badger transaction.
Then, return informing Tendermint that the application is ready to receive application transactions:

```go
func (app *KVStoreApplication) BeginBlock(req abcitypes.RequestBeginBlock) abcitypes.ResponseBeginBlock {
	app.onGoingBlock = app.db.NewTransaction(true)
	return abcitypes.ResponseBeginBlock{}
}
```

Next, let's modify `DeliverTx` to add the `key` and `value` to the database transaction every time our application
receives a new application transaction through `RequestDeliverTx`.

```go
func (app *KVStoreApplication) DeliverTx(req abcitypes.RequestDeliverTx) abcitypes.ResponseDeliverTx {
	if code := app.isValid(req.Tx); code != 0 {
		return abcitypes.ResponseDeliverTx{Code: code}
	}

	parts := bytes.SplitN(req.Tx, []byte("="), 2)
	key, value := parts[0], parts[1]

	if err := app.onGoingBlock.Set(key, value); err != nil {

> 		log.Panicf("Error reading database, unable to verify tx: %v", err)

Reviewer: Please double-check the word "verify"

Author: ```suggestion
		log.Panicf("Error writing to database, unable to execute tx: %v", err)
```

```go
import (
	"bytes"
	"log"

	"github.com/dgraph-io/badger/v3"
	abcitypes "github.com/tendermint/tendermint/abci/types"
)
```

You may have noticed that the application we are writing will crash if it receives
an unexpected error from the Badger database during the `DeliverTx` or `Commit` methods.
This is not an accident. If the application received an error from the database, there 
is no deterministic way for it to make progress so the only safe option is to terminate.

### 1.3.4 Query
When a client tries to read some information from the `kvstore`, the request will be
handled in the `Query` method. To do this, let's rewrite the `Query` method in `app.go`:

```go
func (app *KVStoreApplication) Query(req abcitypes.RequestQuery) abcitypes.ResponseQuery {
	resp := abcitypes.ResponseQuery{Key: req.Data}

	dbErr := app.db.View(func(txn *badger.Txn) error {
		item, err := txn.Get(req.Data)
		if err != nil {
			if err != badger.ErrKeyNotFound {
				return err
			}
			resp.Log = "key does not exist"
			return nil
		}

		return item.Value(func(val []byte) error {
			resp.Log = "exists"
			resp.Value = val
			return nil
		})
	})
	if dbErr != nil {

> 		log.Panicf("Error reading database, unable to verify tx: %v", dbErr)

Reviewer: ```suggestion
		log.Panicf("Error reading database, unable to execute query: %v", dbErr)
```

Reviewer: or something similar?

	}
	return resp
}
```

Since it reads only committed data from the store, transactions that are part of a block

> that is still processed are not reflected in the query result.

Reviewer: ```suggestion
that is being processed are not reflected in the query result.
```

### 1.3.5 PrepareProposal and ProcessProposal

> `PrepareProposal` and `ProcessProposal` are methods introduced in Tendermint 3.7.0 

Reviewer: ```suggestion
`PrepareProposal` and `ProcessProposal` are methods introduced in Tendermint v0.37.0 
```

to give the application more control over the construction and processing of transaction blocks.

When Tendermint Core sees that valid transactions (validated through `CheckTx`) are available to be
included in blocks, it groups some of these transactions and then gives the application a chance 
to modify the group by invoking `PrepareProposal`.

The application is free to modify the group before returning from the call.

> For example, the application may reorder or even remove transactions from the group to improve the

Reviewer: ```suggestion
For example, the application may reorder, add, or even remove transactions from the group to improve the
```

execution of the block once accepted.
In the following code, the application simply returns the unmodified group of transactions:

```go
func (app *KVStoreApplication) PrepareProposal(proposal abcitypes.RequestPrepareProposal) abcitypes.ResponsePrepareProposal {
	return abcitypes.ResponsePrepareProposal{Txs: proposal.Txs}
}
```

Once a proposed block is received by a node, the proposal is passed to the application to give
its blessing before voting to accept the proposal.

This mechanism may be used for different reasons, for example to deal with blocks manipulated 
by malicious nodes, in which case the block should not be considered valid.
The following code simply accepts all proposals:

```go
func (app *KVStoreApplication) ProcessProposal(proposal abcitypes.RequestProcessProposal) abcitypes.ResponseProcessProposal {
	return abcitypes.ResponseProcessProposal{Status: abcitypes.ResponseProcessProposal_ACCEPT}
}
```

> ## 1.4 Starting an application and a Tendermint Core instances

Reviewer: ```suggestion
## 1.4 Starting an application and a Tendermint Core instance
```

> Now that we have the basic functionality of our application in place, let's put it all together inside of our main.go file.

Reviewer: ```suggestion
Now that we have the basic functionality of our application in place, let's put it all together inside of our `main.go` file.
```

> Change the contents of your`main.go` file to the following.

Reviewer: ```suggestion
Change the contents of your `main.go` file to the following.
```

I[2022-11-09|09:08:52.152] received complete proposal block             module=consensus height=2 hash=F518444C0E348270436A73FD0F0B9DFEA758286BEB29482F1E3BEA75330E825C
I[2022-11-09|09:08:52.160] finalizing commit of block                   module=consensus height=2 hash=F518444C0E348270436A73FD0F0B9DFEA758286BEB29482F1E3BEA75330E825C root= num_txs=0
I[2022-11-09|09:08:52.167] executed block                               module=state height=2 num_valid_txs=0 num_invalid_txs=0
I[2022-11-09|09:08:52.171] committed state                              module=state height=2 num_txs=0 app_hash=
```

The blocks, as you can see from the `num_valid_txs=0` part, are empty, but let's remedy that next.

## 1.6 Using the application

Let's try submitting a transaction to our new application.
Open another terminal window and run the following curl command:

```bash
curl -s 'localhost:26657/broadcast_tx_commit?tx="tendermint=rocks"'
```
If everything went well, you should see a response indicating which height the 

> +transaction was included in the blockchain.

Reviewer: ```suggestion
transaction was included in the blockchain.
```

Finally, let's make sure that transaction really was persisted by the application.
Run the following command:

```bash
curl -s 'localhost:26657/abci_query?data="tendermint"'
```

Let's examine the response object that this request returns.
The request returns a `json` object with a `key` and `value` field set.

```json
...
	"key": "dGVuZGVybWludA==",
	"value": "cm9ja3M=",
...
```

Those values don't look like the `key` and `value` we sent to Tendermint.
What's going on here? 

> The response contain a `base64` encoded representation of the data we submitted.

Reviewer: ```suggestion
The response contains a `base64` encoded representation of the data we submitted.
```
